### PR TITLE
update memory index name and add updated_time

### PR DESCRIPTION
**Pull request by Zhangxunmt**
### Description
1. Updated the memory indices names. 
2. Added updated_time field in conversation index. 
3. Update conversation updated_time after created an interaction.
4. Refresh before updating conversations/interactions.
5. Backport https://github.com/opensearch-project/ml-commons/pull/1714 into main. The get interactions in the memory manager is called by chatAgent to list the most recent N interactions from oldest to newest. This is different than the get interactions API required by the OSD FE.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: memory/src/main/java/org/opensearch/ml/memory/index/ConversationMetaIndex.java [hunk 1]
```diff
@@ -366,7 +366,12 @@ public void updateConversation(UpdateRequest updateRequest, ActionListener<Updat
         }
         try (ThreadContext.StoredContext threadContext = client.threadPool().getThreadContext().stashContext()) {
             ActionListener<UpdateResponse> internalListener = ActionListener.runBefore(listener, () -> threadContext.restore());
-            client.update(updateRequest, internalListener);
+            client.admin().indices().refresh(Requests.refreshRequest(META_INDEX_NAME), ActionListener.wrap(refreshResponse -> {
```
Comment (ylwu-amzn) [inline memory/src/main/java/org/opensearch/ml/memory/index/ConversationMetaIndex.java:369]: Why refresh first ? 

Suggest add refresh policy for `updateRequest` 
```
updateRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
```

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/MLMemoryManager.java [hunk 2]
```diff
@@ -106,24 +129,75 @@ public void createInteraction(
...
      */
     public void getFinalInteractions(String conversationId, int lastNInteraction, ActionListener<List<Interaction>> actionListener) {
-        Preconditions.checkNotNull(conversationId);
         Preconditions.checkArgument(lastNInteraction > 0, "lastN must be at least 1.");
-        log.debug("Getting Interactions, conversationId {}, lastN {}", conversationId, lastNInteraction);
-
-        ActionListener<GetInteractionsResponse> al = ActionListener.wrap(getInteractionsResponse -> {
-            actionListener.onResponse(getInteractionsResponse.getInteractions());
-        }, e -> { actionListener.onFailure(e); });
+        log.info("Getting Interactions, conversationId {}, lastN {}", conversationId, lastNInteraction);
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/MLMemoryManager.java:139]: Change it to debug level

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/MLMemoryManager.java [hunk 3]
```diff
@@ -106,24 +129,75 @@ public void createInteraction(
...
+        try (ThreadContext.StoredContext threadContext = client.threadPool().getThreadContext().stashContext()) {
+            ActionListener<List<Interaction>> internalListener = ActionListener.runBefore(listener, () -> threadContext.restore());
+            ActionListener<SearchResponse> al = ActionListener.wrap(response -> {
+                List<Interaction> result = new LinkedList<Interaction>();
+                for (SearchHit hit : response.getHits()) {
+                    result.add(0, Interaction.fromSearchHit(hit));
+                }
+                internalListener.onResponse(result);
+            }, e -> { internalListener.onFailure(e); });
+            client.admin().indices().refresh(Requests.refreshRequest(INTERACTIONS_INDEX_NAME), ActionListener.wrap(r -> {
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/MLMemoryManager.java:196]: If we refresh immediately for write operation(create/update), we don't need to refresh first before searching. 
https://github.com/opensearch-project/ml-commons/pull/1784/files#r1430778796

File: memory/src/main/java/org/opensearch/ml/memory/index/ConversationMetaIndex.java (hunk 1 above)
Comment (Zhangxunmt) [reply memory/src/main/java/org/opensearch/ml/memory/index/ConversationMetaIndex.java]: Removed this one but the refresh is almost everywhere in the Memory layer by Aryn. We should clean them up if they affect the latency.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/MLMemoryManager.java [hunk 4]
```diff
@@ -106,24 +129,75 @@ public void createInteraction(
...
+            ActionListener<Boolean> accessListener = ActionListener.wrap(access -> {
+                if (access) {
+                    innerGetFinalInteractions(conversationId, lastNInteraction, actionListener);
+                } else {
+                    String userstr = client
+                        .threadPool()
+                        .getThreadContext()
+                        .getTransient(ConfigConstants.OPENSEARCH_SECURITY_USER_INFO_THREAD_CONTEXT);
+                    String user = User.parse(userstr) == null ? ActionConstants.DEFAULT_USERNAME_FOR_ERRORS : User.parse(userstr).getName();
+                    throw new OpenSearchSecurityException("User [" + user + "] does not have access to conversation " + conversationId);
```
[the remark this answers is not in the document]
Comment (Zhangxunmt) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/MLMemoryManager.java]: Updated so.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/MLMemoryManager.java (hunk 3 above)
Comment (Zhangxunmt) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/MLMemoryManager.java]: removed this refresh as well.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/MLMemoryManager.java [hunk 5]
```diff
@@ -106,24 +126,68 @@ public void createInteraction(
...
+        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().newStoredContext(true)) {
+            if (!clusterService.state().metadata().hasIndex(INTERACTIONS_INDEX_NAME)) {
+                actionListener.onResponse(List.of());
+                return;
+            }
+            ActionListener<Boolean> accessListener = ActionListener.wrap(access -> {
+                if (access) {
+                    innerGetFinalInteractions(conversationId, lastNInteraction, actionListener);
+                } else {
+                    throw new OpenSearchSecurityException("User does not have access to conversation " + conversationId);
```
Comment (Hailong-am) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/MLMemoryManager.java:147]: should we log the user name?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/MLMemoryManager.java [hunk 6]
```diff
@@ -106,24 +126,68 @@ public void createInteraction(
...
+                }
+            }, e -> { actionListener.onFailure(e); });
+            conversationMetaIndex.checkAccess(conversationId, accessListener);
+        } catch (Exception e) {
+            actionListener.onFailure(e);
```
Comment (Hailong-am) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/MLMemoryManager.java:152]: log the exception?

File: memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java [hunk 7]
```diff
@@ -108,4 +111,34 @@ protected void doExecute(Task task, CreateInteractionRequest request, ActionList
...
 
+    private ActionListener<UpdateResponse> getUpdateResponseListener(
+        String conversationId,
+        String interactionId,
+        ActionListener<CreateInteractionResponse> actionListener
+    ) {
+        return ActionListener.wrap(updateResponse -> {
+            if (updateResponse != null && updateResponse.getResult() == DocWriteResponse.Result.UPDATED) {
+                log
+                    .info(
```
Comment (ylwu-amzn) [inline memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java:122]: change to debug level?

**Review by Hailong-am — APPROVED** [review]

**Review by ylwu-amzn — APPROVED** [review]

File: memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java [hunk 8]
```diff
@@ -93,10 +96,9 @@ protected void doExecute(Task task, CreateInteractionRequest request, ActionList
         Integer traceNumber = request.getTraceNumber();
         try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().newStoredContext(true)) {
             ActionListener<CreateInteractionResponse> internalListener = ActionListener.runBefore(actionListener, () -> context.restore());
-            ActionListener<String> al = ActionListener
-                .wrap(iid -> { internalListener.onResponse(new CreateInteractionResponse(iid)); }, e -> {
-                    internalListener.onFailure(e);
-                });
+            ActionListener<String> al = ActionListener.wrap(iid -> {
```
Comment (dhrubo-os) [inline memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java:99]: Don't we need to stash the context before creating the interaction? Does `newStoredContext` stashes the context?

File: memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java [hunk 9]
```diff
@@ -93,10 +96,9 @@ protected void doExecute(Task task, CreateInteractionRequest request, ActionList
...
+                cmHandler.updateConversation(cid, new HashMap<>(), getUpdateResponseListener(cid, iid, internalListener));
+            }, e -> { internalListener.onFailure(e); });
             if (parintIid == null || traceNumber == null) {
                 cmHandler.createInteraction(cid, inp, prompt, rsp, ogn, additionalInfo, al);
```
Comment (dhrubo-os) [inline memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java:103]: [nit] `cid`, `inp`, `rsp`, `ogn` is super hard to understand.

File: memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java [hunk 10]
```diff
@@ -93,10 +96,9 @@ protected void doExecute(Task task, CreateInteractionRequest request, ActionList
...
         try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().newStoredContext(true)) {
             ActionListener<CreateInteractionResponse> internalListener = ActionListener.runBefore(actionListener, () -> context.restore());
-            ActionListener<String> al = ActionListener
-                .wrap(iid -> { internalListener.onResponse(new CreateInteractionResponse(iid)); }, e -> {
-                    internalListener.onFailure(e);
-                });
+            ActionListener<String> al = ActionListener.wrap(iid -> {
+                cmHandler.updateConversation(cid, new HashMap<>(), getUpdateResponseListener(cid, iid, internalListener));
+            }, e -> { internalListener.onFailure(e); });
             if (parintIid == null || traceNumber == null) {
```
Comment (dhrubo-os) [inline memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java:102]: [nit] what is `parintIid`? I guess this is parent id of something. Parent id of what?

File: memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java [hunk 11]
```diff
@@ -108,4 +110,34 @@ protected void doExecute(Task task, CreateInteractionRequest request, ActionList
...
+    ) {
+        return ActionListener.wrap(updateResponse -> {
+            if (updateResponse != null && updateResponse.getResult() == DocWriteResponse.Result.UPDATED) {
+                log
+                    .debug(
+                        "Successfully updated the Conversation with ID: {} after interaction {} is created",
+                        conversationId,
+                        interactionId
+                    );
+                actionListener.onResponse(new CreateInteractionResponse(interactionId));
```
Comment (dhrubo-os) [inline memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java:126]: [nit] The same line repeated 3 times in the same method. May be we can simplify the code?

File: memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java (hunk 8 above)
Comment (Zhangxunmt) [reply memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java:99]: Not here. If you trace the code, stash is done in the InteractionsIndex through ConversationalMemoryHandler

File: memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java (hunk 9 above)
Comment (Zhangxunmt) [reply memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java:103]: can update in a separate PR. These names are consistently used in the repo.

File: memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java (hunk 10 above)
Comment (Zhangxunmt) [reply memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java:102]: parent Id is for traces. parent id is null for final interactions. traces use this parent id to find the final interaction.

File: memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java (hunk 11 above)
Comment (Zhangxunmt) [reply memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java:126]: They are not redundant. They are for different conditions in each branch. This onResponse is used frequently everywhere to send response back to users.

**Review by b4sjoo — APPROVED** [review]

File: memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java (hunk 11 above)
Comment (dhrubo-os) [reply memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java:126]: I don't understand how `actionListener.onResponse(new CreateInteractionResponse(interactionId));` is different in each branch?

Comment (Zhangxunmt) [reply memory/src/main/java/org/opensearch/ml/memory/action/conversation/CreateInteractionTransportAction.java:126]: They are the same, but each is needed in each branch. The two of them can be moved outside of the if ... else ... and replaced by a single line, but essentially nothing changes. The actionListener.onResponse is required in the error catch anyways.

**Pull request merged**